datastore/postgres/deletemanifests.go

```diff
@@ -33,55 +36,83 @@ var (
 
 func (s *IndexerStore) DeleteManifests(ctx context.Context, d ...claircore.Digest) ([]claircore.Digest, error) {
 	ctx = zlog.ContextWithValues(ctx, "component", "datastore/postgres/DeleteManifests")
-	rm, err := s.deleteManifests(ctx, d)
-	if err != nil {
-		return nil, err
-	}
-	return rm, s.layerCleanup(ctx)
+	return s.deleteManifests(ctx, d)
 }
 
-func (s *IndexerStore) deleteManifests(ctx context.Context, d []claircore.Digest) ([]claircore.Digest, error) {
-	const deleteManifest = `DELETE FROM manifest WHERE hash = ANY($1::TEXT[]) RETURNING manifest.hash;`
-	var err error
-	defer promTimer(deleteManifestsDuration, "deleteManifest", &err)()
-	defer func(e *error) {
-		deleteManifestsCounter.WithLabelValues("deleteManifest", success(*e)).Inc()
-	}(&err)
-	rows, err := s.pool.Query(ctx, deleteManifest, digestSlice(d))
-	if err != nil {
-		return nil, err
-	}
-	defer rows.Close()
-	rm := make([]claircore.Digest, 0, len(d)) // May over-allocate, but at least it's only doing it once.
-	for rows.Next() {
-		i := len(rm)
-		rm = rm[:i+1]
-		err = rows.Scan(&rm[i])
-		if err != nil {
-			return nil, err
-		}
-	}
-	err = rows.Err()
-	if err != nil {
-		return nil, err
-	}
-	zlog.Debug(ctx).
-		Int("count", len(rm)).
-		Int("nonexistant", len(d)-len(rm)).
-		Msg("deleted manifests")
-	return rm, nil
-}
+func (s *IndexerStore) deleteManifests(ctx context.Context, ds []claircore.Digest) ([]claircore.Digest, error) {
+	const (
+		getManifestID      = `SELECT id FROM manifest WHERE hash = $1`
+		getLayers          = `SELECT layer_id FROM manifest_layer WHERE manifest_id = $1;`
+		getDeletableLayers = `
+SELECT l.id FROM layer l 
+LEFT JOIN manifest_layer ml 
+ON l.id = ml.layer_id 
+WHERE l.id = $1
+AND ml.layer_id IS NULL;`
+		deleteManifest = `DELETE FROM manifest WHERE id = $1;`
+		deleteLayers   = `DELETE FROM layer WHERE id = ANY($1);`
+	)
+
+	deletedManifests := make([]claircore.Digest, 0, len(ds))
+	for _, d := range ds {
+		s.pool.BeginFunc(ctx, func(tx pgx.Tx) error {
```

Comment:
What's the thinking on doing a transaction per loop, rather than a loop in a transaction?

Reply:
My theory was that each manifest deletion can be logically broken down into one transaction's worth of work and it makes the code slightly cleaner when trying not to stack up deferred `Close()`s.